src/main/java/io/wgj/ctci/Stack.java

```diff
@@ -1,21 +1,80 @@
 package io.wgj.ctci;
 
-public class Stack {
-    int[] stack;
+/**
+ * Stack is a stack data structure backed by a dynamic array. Stack can be
+ * initialized with a starting array size. It's written to develop familiarity
+ * with the stack data structure, and should otherwise not be used.
+ * Maximum allowed array is 1024. If you set a starting capacity larger than
+ * this, the maximum allowed array size is used as the starting capacity. An
+ * UnsupportedOperationException will be thrown when more the maximum allowed
+ * amount of elements are added.
+ */
+public class Stack<E> {
+    static final int MAX_ARRAY_SIZE = 1024;
+    static final int DEFAULT_CAPACITY = 10;
+    Object[] stack;
+    int size;
     int position;
 
-    public Stack(){
-        stack = new int[10];
+    public Stack() {
+        this(DEFAULT_CAPACITY);
+    }
+
+    public Stack(int size) {
+        stack = new Object[size];
```

Comment:
> Setting this before checking constraints on size means your array could be bigger than necessary. 

Not sure what you mean. In the case where the caller creates a `Stack` without using it, an array size of 10 seems like small enough waste to not be terrible mistake. In the case where the caller sets an initial capacity, and they don't use it seems like a mistake, but I think that's the responsibility of the caller. Guarding against that seems like it would be at the expense of responsible users, wanting to avoid unnecessary growths.

> why have a maximum size at all?

To prevent oom errors. Granted `1024` is too low, but it's gotta be _something_ and it's a toy program. I wanted to promote responsible array use, but I'd imagine setting the maximum size as `Integer.MAX_VALUE` would better show that intent.

> Is it part of the exercise?

No.

> Otherwise it seem like it just makes this class less usable and more complicated.

Agreed. The goal isn't generally usability by another programmer, but familiarity with the stack and dynamic array data structure. With that in mind, I chose to not use unbounded arrays, and to not short change myself on the last growth operation. I thought it was more "right" to solve for those constraints than to pass the buck to the calling programmer.

Reply:
Per our conversation, it's the caller's responsibility to prevent overflow of the data structure.